cmd/find/timeAttrPrimary.go

```diff
@@ -0,0 +1,158 @@
+	startTime = time
+}
+
+// Use durationOf instead of a hash so that we generate a more readable
+// panic message
+func durationOf(unit byte) time.Duration {
+	switch unit {
+	case 's':
+		return time.Second
+	case 'm':
+		return time.Minute
+	case 'h':
+		return time.Hour
+	case 'd':
+		return 24 * time.Hour
+	case 'w':
+		return 7 * 24 * time.Hour
+	default:
+		panic(fmt.Sprintf("cmdfind.durationOf received an unexpected unit %v", unit))
+	}
+}
+
+// We use getTimeAttrValue to retrieve the time attribute's value for performance
+// reasons. Using e.Attributes.ToMap()[name] would be slower because it would
+// require an additional type assertion to extract the time.Time object.
+func getTimeAttrValue(name string, e *apitypes.ListEntry) (time.Time, bool) {
+	switch name {
+	case "ctime":
+		return e.Attributes.Ctime(), e.Attributes.HasCtime()
+	case "mtime":
+		return e.Attributes.Mtime(), e.Attributes.HasMtime()
+	case "atime":
+		return e.Attributes.Atime(), e.Attributes.HasAtime()
+	default:
+		panic(fmt.Sprintf("cmdfind.getTimeAttrValue called with nonexistent time attribute %v", name))
+	}
+}
+
+var timeAttrValueRegex = regexp.MustCompile(`^(\+|-)?((\d+)|(\d+[smhdw])+)$`)
+
+// Describes values like 1h30m, 3w2d, etc.
+var timeAttrUnitValueRegex = regexp.MustCompile(`\d+[smhdw]`)
+
+// parseDuration is a helper for newTimeAttrPrimary that parses the
+// duration from v. It returns the parsed duration, and a boolean
+// indicating whether the difference between the entry's time attribute
+// and start time needs to be rounded to the next 24-hour period (which
+// is only true if v is an integer).
+func parseDuration(v string) (time.Duration, bool) {
+	var roundDiff bool
+	var duration time.Duration
+	if n, err := strconv.ParseInt(v, 10, 32); err == nil {
+		// v (n) is an integer
+		duration = time.Duration(n) * durationOf('d')
+		roundDiff = true
+	} else {
+		// v consists of individual time units. Add them up to get the overall
+		// duration. Note that repeated units will be stacked, meaning something
+		// like "1h1h1h1h" will result in a duration of 4 hours. This behavior matches
+		// BSD's find command's behavior.
+		for _, chunk := range timeAttrUnitValueRegex.FindAllString(v, -1) {
+			endIx := len(chunk) - 1
+			unit := chunk[endIx]
+			n, err := strconv.ParseInt(chunk[0:endIx], 10, 32)
+			if err != nil {
+				// We should never hit this code-path because timeAttrUnitValueRegex
+				// already verified that chunk[0:endIx] is an integer.
+				msg := fmt.Sprintf("errored parsing duration %v, which is an expected integer: %v", chunk[0:endIx], err)
+				panic(msg)
+			}
+			duration += time.Duration(n) * durationOf(unit)
+		}
+	}
+	return duration, roundDiff
+}
+
+// timeAttrPrimary => -<name> (+|-)?((\d+)|(\d+[smhdw])
+//
+// where s => seconds, m => minutes, h => hours, d => days, w => weeks
+//
+// Example inputs:
+//   -mtime 1      (true if the difference between the entry's mtime and startTime is exactly 1 24-hour period)
```

Comment:
Is there any rounding on that? Alternatively, is there any way to express "between 23 and 24 hours"? I guess that could be `-mtime +23h -and -mtime -24h`.

Reply:
Yeah, that seems to be the only way to do it. We could consider adding alternate syntax for these primaries that supports something like that? Wouldn't be hard to do if people find `and`'ing the `mtimes` annoying.

Reply:
I think the way I outlined is fine. Not sure exact numbers will be that useful in practice.